File: Employee.java
public class Employee extends User {
    int managerID;
    Boolean isAdmin;

    //Constructors
    public Employee(String newEmail) {
        super(newEmail);
    }

    public Employee(String newEmail, String newFirstName, String newLastName) {
        super(newEmail, newFirstName, newLastName);
    }
    
    public Employee(int newID, String newEmail, String newFirstName, String newLastName) {
        super(newID, newEmail, newFirstName, newLastName);
    }

    public Employee(int newID, String newEmail, String newFirstName, String newLastName, Boolean newEmployeeIsAdmin) {
        super(newID, newEmail, newFirstName, newLastName);
        this.isAdmin = newEmployeeIsAdmin;
    }

    public Employee(String newEmail, String newFirstName, String newLastName, Boolean newEmployeeIsAdmin) {
        super(newEmail, newFirstName, newLastName);
        this.isAdmin = newEmployeeIsAdmin;
    }

    //Getters/setters

    public void setManagerID(int newManagerID) {
        this.managerID = newManagerID;
    }

    public int getManagerID() {
        return this.managerID;
    }

    public void setIsAdmin(boolean updatedIsAdmin) {
        this.isAdmin = updatedIsAdmin;
    }

    public boolean isAdmin() {
        return this.isAdmin;
    }

}

File: models/Employee.js
const User = require('./User'); // Import User class
const pool = require('../config/db'); // Import database connection

class Employee extends User {
    constructor(id, email, firstName, lastName, role = 'employee', isAdmin = false, managerID = null) {
        super(id, email, firstName, lastName, role); // Call User constructor
        this.isAdmin = isAdmin;
        this.managerID = managerID;
    }

    // Getter and Setter for Manager ID
    getManagerID() {
        return this.managerID;
    }

    setManagerID(managerID) {
        this.managerID = managerID;
    }

    // Getter and Setter for isAdmin
    getIsAdmin() {
        return this.isAdmin;
    }

    setIsAdmin(isAdmin) {
        this.isAdmin = isAdmin;
    }

    // Save employee to the database
    async save() {
        // Save as a User first
        if (!this.id) {
            await super.save(); // Use User's save method to generate the user ID
        }

        const sql = `
            INSERT INTO Employees (userID, isAdmin, managerID)
            VALUES (?, ?, ?)
            ON DUPLICATE KEY UPDATE isAdmin = VALUES(isAdmin), managerID = VALUES(managerID)
        `;
        await pool.query(sql, [this.id, this.isAdmin, this.managerID]);
    }

    // Fetch an employee by user ID
    static async getByUserID(userID) {
        const sql = `
            SELECT u.id, u.email, u.firstName, u.lastName, u.role, e.isAdmin, e.managerID
            FROM Users u
            LEFT JOIN Employees e ON u.id = e.userID
            WHERE u.id = ?
        `;
        const [rows] = await pool.query(sql, [userID]);
        if (rows.length === 0) {
            return null;
        }

        const { id, email, firstName, lastName, role, isAdmin, managerID } = rows[0];
        return new Employee(id, email, firstName, lastName, role, isAdmin, managerID);
    }

    // Assign a task to this employee
    async assignTask(taskID) {
        const sql = `UPDATE Tasks SET assignedTo = ? WHERE id = ?`;
        await pool.query(sql, [this.id, taskID]);
    }
}

module.exports = Employee;
